### PR TITLE
Give LineEdit/TextEdit a custom color for font while uneditable

### LINKED ISSUE #29814: LineEdit in mode: editable = false, - translucent text
**Godot version: 3.1.1**

I need not editable LineEdit:

`MyLineEdit.editable = false`

And I see that the text in LineEdit is translucent in this mode. This is not what was expected. But most importantly, I didn’t find a parameter to change it.

### PR DESCRIPTION
**Pull request by GlaceGwyneth**
By default an uneditable LineEdit's text has some transparency hardcoded.
This change adds a custom color to LineEdit for adjusting the font when editable is unchecked.
It also leaves the default behavior (0.5 alpha) unchanged.

Fixes #29814

### CODE REVIEW

**Comment by YeldhamDev** [comment]
I believe `TextEdit` also does that when in read-only mode. This change should be made there as well then.

**Comment by GlaceGwyneth** [comment]
Hopefully I didn't miss anything while testing...
One thing I noticed:
The order of the custom colors in editor isn't the same as the order in the code, so the readonly color is way off at the bottom by itself, but I'm not sure how to control this?

**Comment by YeldhamDev** [comment]
Thank you for the changes. And happy first contribution!

**Comment by YeldhamDev** [comment]
From my quick look in the editor, it seems that code order doesn't affect constants, so you're in the green on that.

However, Travis detect some errors. I will comment on the lines that they were found shortly.

File: scene/gui/text_edit.cpp [hunk 1]
```diff
@@ -675,8 +675,8 @@ void TextEdit::_notification(int p_what) {
 
 			int visible_rows = get_visible_rows() + 1;
 
-			Color color = cache.font_color;
-			color.a *= readonly_alpha;
+			Color color = readonly ? cache.font_color_readonly : cache.font_color;
+
```
Comment (YeldhamDev) [inline scene/gui/text_edit.cpp]: Unneeded extra line.

**Comment by YeldhamDev** [comment]
The `readonly_alpha` variable found in line 665 is now unused and should be removed.
(which I would have loved to have commented directly on it, but GitHub doesn't allow commenting on lines too far away from modifications because "reasons" and- this is not the proper place to rant about that isn't it?)

**Comment by YeldhamDev** [comment]
That should be it. I recommend getting Godot's [git hooks](https://github.com/godotengine/godot/tree/master/misc/hooks), so clang will alert you when breaking formatting.

File: scene/resources/default_theme/default_theme.cpp [hunk 2]
```diff
@@ -440,6 +440,7 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
 	theme->set_color("completion_font_color", "TextEdit", Color::html("aaaaaa"));
 	theme->set_color("font_color", "TextEdit", control_font_color);
 	theme->set_color("font_color_selected", "TextEdit", Color(0, 0, 0));
+	theme->set_color("font_color_read_only", "TextEdit", Color(control_font_color.r, control_font_color.g, control_font_color.b, 0.5f));
```
Comment (akien-mga) [inline scene/resources/default_theme/default_theme.cpp]: The property is named `readonly` in one word, so it's probably best if the color follows the same spelling.

File: scene/gui/line_edit.cpp [hunk 3]
```diff
@@ -724,15 +724,14 @@ void LineEdit::_notification(int p_what) {
...
-			Color font_color = get_color("font_color");
+			Color font_color = is_editable() ? get_color("font_color") : get_color("font_color_uneditable");
 			Color font_color_selected = get_color("font_color_selected");
 			Color cursor_color = get_color("cursor_color");
 
 			const String &t = using_placeholder ? placeholder : text;
 			// draw placeholder color
 			if (using_placeholder)
 				font_color.a *= placeholder_alpha;
-			font_color.a *= disabled_alpha;
```
Comment (akien-mga) [inline scene/gui/line_edit.cpp]: I think you can remove the `disabled_alpha` variable which is now only used once at line 740 (you can use a ternary operator there instead to set the alpha).

**Review by akien-mga — APPROVED** [review]
Looks good, and it seems like you made sure the new properties will produce the same visual effect by default, so it should not break compatibility.

I left a couple comments, otherwise it should be good to merge.

File: scene/gui/line_edit.cpp (hunk 3 above)
Comment (GlaceGwyneth) [reply scene/gui/line_edit.cpp]: I felt a bit weird about the ternary bit, does it look okay?

Comment (akien-mga) [reply scene/gui/line_edit.cpp]: I would have simplified to `is_editable() ? 0.9 : 0.45`, but it's also ok as is.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Shadowblitz16** [comment]
@akien-mga was this reverted?
there is no option for disabled text color for line edit
image

**Comment by akien-mga** [comment]
I see "Read Only" among colors in your screenshot...

**Comment by Shadowblitz16** [comment]
@akien-mga  the thing is its not changing anything

**Comment by bruvzg** [comment]
Color name is `font_color_uneditable`, and it's available if you add the item to the theme `Edit Theme->Add Item`, select `Type: LineEdit`, `Data Type: Color`, `Name: font_color_uneditable`. There's no color with name `read_only`, you are probably using a theme created in the old engine version.

**Pull request merged**